src/application/transaction/verifyTx.ts

```diff
@@ -0,0 +1,302 @@
+              }),
+            )
+          }
+          case SubStateType.PREPARED_STAKE:
+          case SubStateType.PREPARED_UNSTAKE:
+          case SubStateType.STAKE_OWNERSHIP: {
+            return accountAddressFromREAddressAndNetwork({
+              reAddress: substate.owner,
+              network,
+            }).map(
+              (owner): PreparedStake | PreparedUnstake | StakeOwnership => ({
+                amount: substate.amount,
+                owner,
+                validator: ValidatorAddress.fromPublicKeyAndNetwork({
+                  publicKey: substate.validator,
+                  network,
+                }),
+                substateType: substate.substateType,
+              }),
+            )
+          }
+
+          default:
+            return ok(null)
+        }
+      }),
+    ),
+  )
+
+const assert = (condition: boolean, errorMessage: string) =>
+  condition ? ok(null) : err(errorMessage)
+
+const assertSubstateTypes = (
+  expectedSubStateTypes: SubStateType[],
+  subStateTypes: SubStateType[],
+) =>
+  assert(
+    subStateTypes.every(substateType =>
+      expectedSubStateTypes.includes(substateType),
+    ),
+    `incorrect subStateTypes, expected ${expectedSubStateTypes.join()} got '${subStateTypes.join(
+      ' ',
+    )}'`,
+  )
+
+const assertAccountAddress = (
+  type: 'from' | 'to',
+  expected: AccountAddressT | ValidatorAddressT,
+  actual: AccountAddressT | ValidatorAddressT,
+) =>
+  assert(
+    expected.equals(actual),
+    `incorrect ${type} address, expected '${expected.toPrimitive()}' got '${actual.toPrimitive()}'`,
+  )
+
+const assertNumberOfInstructions = (expected: number, actual: number) =>
+  assert(
+    expected === actual,
+    `incorrect number of instructions, expected '${expected}' got '${actual}'`,
+  )
+
+const assertAmount = (expected: AmountT, actual: AmountT) =>
+  assert(
+    actual.eq(expected),
+    `incorrect amount expected '${expected.toString()}' got '${actual.toString()}'`,
+  )
+
+const assertResource = (expected: ResourceIdentifierT, actual: string) => {
+  const expectedHash = expected.hash.toString('hex')
+  return assert(
+    expectedHash === actual,
+    `incorrect rri expected '${expectedHash}' got '${actual}'`,
+  )
+}
+
+const verifyTransfer = (
+  action: IntendedTransferTokensAction,
+  instructions: TokensGroup,
+) => {
+  const [from, to] = instructions
+  const isSelfTransfer = from.owner.equals(to.owner)
+
+  return combineWithAllErrors([
+    assertNumberOfInstructions(2, instructions.length),
+    assertSubstateTypes(
+      [SubStateType.TOKENS],
+      instructions.map(({ substateType }) => substateType),
+    ),
+    assertAccountAddress('from', action.from_account, from.owner),
+    assertAccountAddress('to', action.to_account, to.owner),
+    assertResource(action.rri, from.resource),
+    assertResource(action.rri, to.resource),
+    isSelfTransfer ? ok(null) : assertAmount(action.amount, to.amount),
```

Comment:
this, for some reason, messes up the return type of the `Result`.

fix: 

```suggestion
    isSelfTransfer ? ok<null, string>(null) : assertAmount(action.amount, to.amount),
```